# eBay Kleinanzeigen: Used Car Sales

<br></br>
eBay Kleinanzeigen is the German eBay website. Similar to the US eBay, the site hosts different classified listings. These listings include home and garden items, electronics, collectibles, services, and jobs, to name a few.

The aim of this project is to clean and analyze an eBay Kleinanzeigen dataset using Pandas.

## Dataset: Used Car Listings

The dataset used in this project is a sample of [the original dataset](https://www.kaggle.com/orgesleka/used-cars-database/data), which was scraped and uploaded to Kaggle. The dataset was reduced from the 370,000+ rows to 50,000 entries to ensure the code runs quickly in a hosted environment. The dataset has also been 'dirtied' a bit to more closely resemble real-life, unsanitized scraped datasets.

The data dictionary provided contains the following details:

| COLUMN NAME    | DESCRIPTION  |
|:---------------|:-------------|
| **dateCrawled** | When this ad was first crawled. All field-values are taken from this date. |
| **name** | Name of the car. |
| **seller** | Whether the seller is private or a dealer. |
| **offerType** | The type of listing |
| **price** | The price on the ad to sell the car. |
| **abtest** | Whether the listing is included in an A/B test. |
| **vehicleType** | The vehicle Type. |
| **yearOfRegistration** | The year in which the car was first registered. |
| **gearbox** | The transmission type. |
| **powerPS** | The power of the car in PS. |
| **model** | The car model name. |
| **kilometer** | How many kilometers the car has driven. |
| **monthOfRegistration** | The month in which the car was first registered. |
| **fuelType** | What type of fuel the car uses. |
| **brand** | The brand of the car. |
| **notRepairedDamage** | If the car has a damage which is not yet repaired. |
| **dateCreated** | The date on which the eBay listing was created. |
| **nrOfPictures** | The number of pictures in the ad. |
| **postalCode** | The postal code for the location of the vehicle. |
| **lastSeenOnline** | When the crawler saw this ad last online. |

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('datasets/autos.csv', encoding='Latin-1')
autos_sample = autos[7:12].T
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [2]:
autos_sample

7                    8  \
dateCrawled           2016-03-16 18:55:19  2016-03-22 16:51:34   
name                 Golf_IV_1.9_TDI_90PS           Seat_Arosa   
seller                             privat               privat   
offerType                         Angebot              Angebot   
price                              $1,990                 $250   
abtest                            control                 test   
vehicleType                     limousine                  NaN   
yearOfRegistration                   1998                 2000   
gearbox                           manuell              manuell   
powerPS                                90                    0   
model                                golf                arosa   
odometer                        150,000km            150,000km   
monthOfRegistration                    12                   10   
fuelType                           diesel                  NaN   
brand                          volkswagen                 seat   
notRepairedDamage                    nein                 nein   
dateCreated           2016-03-16 00:00:00  2016-03-22 00:00:00   
nrOfPictures                            0                    0   
postalCode                          53474                 7426   
lastSeen              2016-04-07 03:17:32  2016-03-26 18:18:10   

                                                             9  \
dateCrawled                                2016-03-16 13:47:02   
name                 Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
seller                                                  privat   
offerType                                              Angebot   
price                                                     $590   
abtest                                                 control   
vehicleType                                                bus   
yearOfRegistration                                        1997   
gearbox                                                manuell   
powerPS                                                     90   
model                                                   megane   
odometer                                             150,000km   
monthOfRegistration                                          7   
fuelType                                                benzin   
brand                                                  renault   
notRepairedDamage                                         nein   
dateCreated                                2016-03-16 00:00:00   
nrOfPictures                                                 0   
postalCode                                               15749   
lastSeen                                   2016-04-06 10:46:35   

                                               10                          11  
dateCrawled                   2016-03-15 01:41:36         2016-03-16 18:45:34  
name                 VW_Golf_Tuning_in_siber/grau  Mercedes_A140_Motorschaden  
seller                                     privat                      privat  
offerType                                 Angebot                     Angebot  
price                                        $999                        $350  
abtest                                       test                     control  
vehicleType                                   NaN                         NaN  
yearOfRegistration                           2017                        2000  
gearbox                                   manuell                         NaN  
powerPS                                        90                           0  
model                                         NaN                         NaN  
odometer                                150,000km                   150,000km  
monthOfRegistration                             4                           0  
fuelType                                   benzin                      benzin  
brand                                  volkswagen               me

Some initial details about the data:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values. `notRepairedDamage` has 19.6%.
- The column names use camelcase, which would be converted to the preferred snakecase.

## Data Cleaning

### Updating Column Names

Some of the column names have been renamed from camelcase to snakecase. Some of the column names were also updated to be more descriptive.

In [3]:
rename_dict = {'dateCrawled' : 'date_crawled',
               'offerType' : 'offer_type',
               'vehicleType' : 'vehicle_type',
               'yearOfRegistration' : 'registration_year',
               'powerPS' : 'power_ps',
               'monthOfRegistration' : 'registration_month',
               'fuelType' : 'fuel_type',
               'notRepairedDamage' : 'unrepaired_damage',
               'dateCreated' : 'ad_created',
               'nrOfPictures' : 'nr_of_pictures',
               'postalCode' : 'postal_code',
               'lastSeen' : 'last_seen'
              }

autos.rename(rename_dict, axis=1, inplace=True)
autos_sample

7                    8  \
dateCrawled           2016-03-16 18:55:19  2016-03-22 16:51:34   
name                 Golf_IV_1.9_TDI_90PS           Seat_Arosa   
seller                             privat               privat   
offerType                         Angebot              Angebot   
price                              $1,990                 $250   
abtest                            control                 test   
vehicleType                     limousine                  NaN   
yearOfRegistration                   1998                 2000   
gearbox                           manuell              manuell   
powerPS                                90                    0   
model                                golf                arosa   
odometer                        150,000km            150,000km   
monthOfRegistration                    12                   10   
fuelType                           diesel                  NaN   
brand                          volkswagen                 seat   
notRepairedDamage                    nein                 nein   
dateCreated           2016-03-16 00:00:00  2016-03-22 00:00:00   
nrOfPictures                            0                    0   
postalCode                          53474                 7426   
lastSeen              2016-04-07 03:17:32  2016-03-26 18:18:10   

                                                             9  \
dateCrawled                                2016-03-16 13:47:02   
name                 Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
seller                                                  privat   
offerType                                              Angebot   
price                                                     $590   
abtest                                                 control   
vehicleType                                                bus   
yearOfRegistration                                        1997   
gearbox                                                manuell   
powerPS                                                     90   
model                                                   megane   
odometer                                             150,000km   
monthOfRegistration                                          7   
fuelType                                                benzin   
brand                                                  renault   
notRepairedDamage                                         nein   
dateCreated                                2016-03-16 00:00:00   
nrOfPictures                                                 0   
postalCode                                               15749   
lastSeen                                   2016-04-06 10:46:35   

                                               10                          11  
dateCrawled                   2016-03-15 01:41:36         2016-03-16 18:45:34  
name                 VW_Golf_Tuning_in_siber/grau  Mercedes_A140_Motorschaden  
seller                                     privat                      privat  
offerType                                 Angebot                     Angebot  
price                                        $999                        $350  
abtest                                       test                     control  
vehicleType                                   NaN                         NaN  
yearOfRegistration                           2017                        2000  
gearbox                                   manuell                         NaN  
powerPS                                        90                           0  
model                                         NaN                         NaN  
odometer                                150,000km                   150,000km  
monthOfRegistration                             4                           0  
fuelType                                   benzin                      benzin  
brand                                  volkswagen               me

In [4]:
autos.describe(include='all').T

count unique                  top   freq     mean  \
date_crawled        50000  48213  2016-03-08 10:40:35      3      NaN   
name                50000  38754          Ford_Fiesta     78      NaN   
seller              50000      2               privat  49999      NaN   
offer_type          50000      2              Angebot  49999      NaN   
price               50000   2357                   $0   1421      NaN   
abtest              50000      2                 test  25756      NaN   
vehicle_type        44905      8            limousine  12859      NaN   
registration_year   50000    NaN                  NaN    NaN  2005.07   
gearbox             47320      2              manuell  36993      NaN   
power_ps            50000    NaN                  NaN    NaN  116.356   
model               47242    245                 golf   4024      NaN   
odometer            50000     13            150,000km  32424      NaN   
registration_month  50000    NaN                  NaN    NaN  5.72336   
fuel_type           45518      7               benzin  30107      NaN   
brand               50000     40           volkswagen  10687      NaN   
unrepaired_damage   40171      2                 nein  35232      NaN   
ad_created          50000     76  2016-04-03 00:00:00   1946      NaN   
nr_of_pictures      50000    NaN                  NaN    NaN        0   
postal_code         50000    NaN                  NaN    NaN  50813.6   
last_seen           50000  39481  2016-04-07 06:17:27      8      NaN   

                        std   min    25%    50%    75%    max  
date_crawled            NaN   NaN    NaN    NaN    NaN    NaN  
name                    NaN   NaN    NaN    NaN    NaN    NaN  
seller                  NaN   NaN    NaN    NaN    NaN    NaN  
offer_type              NaN   NaN    NaN    NaN    NaN    NaN  
price                   NaN   NaN    NaN    NaN    NaN    NaN  
abtest                  NaN   NaN    NaN    NaN    NaN    NaN  
vehicle_type            NaN   NaN    NaN    NaN    NaN    NaN  
registration_year   105.713  1000   1999   2003   2008   9999  
gearbox                 NaN   NaN    NaN    NaN    NaN    NaN  
power_ps            209.217     0     70    105    150  17700  
model                   NaN   NaN    NaN    NaN    NaN    NaN  
odometer                NaN   NaN    NaN    NaN    NaN    NaN  
registration_month  3.71198     0      3      6      9     12  
fuel_type               NaN   NaN    NaN    NaN    NaN    NaN  
brand                   NaN   NaN    NaN    NaN    NaN    NaN  
unrepaired_damage       NaN   NaN    NaN    NaN    NaN    NaN  
ad_created              NaN   NaN    NaN    NaN    NaN    NaN  
nr_of_pictures            0     0      0      0      0      0  
postal_code         25779.7  1067  30451  49577  71540  99998  
last_seen               NaN   NaN    NaN    NaN    NaN    NaN

To prepare the data for analysis, there are other cleaning tasks that need to be completed. They are:
- Drop columns where all or almost all values are the same, since they won't add anything to analysis
- Clean and convert numeric data stored as text

### Dropping Text Columns

The following columns will be deleted:
- `seller` - almost all values are `privat`
- `offer_type` - almost all values are `Angebot`
- `nr_of_pictures` - almost all values are `0`

In [5]:
dropped_columns = ['seller', 'offer_type', 'nr_of_pictures']

for column in dropped_columns:
    print(autos[column].value_counts())
    if column != dropped_columns[-1]:
        print('\n')

privat        49999
gewerblich        1
Name: seller, dtype: int64


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


0    50000
Name: nr_of_pictures, dtype: int64


In [6]:
autos.drop(dropped_columns, axis=1, inplace=True)
autos_sample

7                    8  \
dateCrawled           2016-03-16 18:55:19  2016-03-22 16:51:34   
name                 Golf_IV_1.9_TDI_90PS           Seat_Arosa   
seller                             privat               privat   
offerType                         Angebot              Angebot   
price                              $1,990                 $250   
abtest                            control                 test   
vehicleType                     limousine                  NaN   
yearOfRegistration                   1998                 2000   
gearbox                           manuell              manuell   
powerPS                                90                    0   
model                                golf                arosa   
odometer                        150,000km            150,000km   
monthOfRegistration                    12                   10   
fuelType                           diesel                  NaN   
brand                          volkswagen                 seat   
notRepairedDamage                    nein                 nein   
dateCreated           2016-03-16 00:00:00  2016-03-22 00:00:00   
nrOfPictures                            0                    0   
postalCode                          53474                 7426   
lastSeen              2016-04-07 03:17:32  2016-03-26 18:18:10   

                                                             9  \
dateCrawled                                2016-03-16 13:47:02   
name                 Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
seller                                                  privat   
offerType                                              Angebot   
price                                                     $590   
abtest                                                 control   
vehicleType                                                bus   
yearOfRegistration                                        1997   
gearbox                                                manuell   
powerPS                                                     90   
model                                                   megane   
odometer                                             150,000km   
monthOfRegistration                                          7   
fuelType                                                benzin   
brand                                                  renault   
notRepairedDamage                                         nein   
dateCreated                                2016-03-16 00:00:00   
nrOfPictures                                                 0   
postalCode                                               15749   
lastSeen                                   2016-04-06 10:46:35   

                                               10                          11  
dateCrawled                   2016-03-15 01:41:36         2016-03-16 18:45:34  
name                 VW_Golf_Tuning_in_siber/grau  Mercedes_A140_Motorschaden  
seller                                     privat                      privat  
offerType                                 Angebot                     Angebot  
price                                        $999                        $350  
abtest                                       test                     control  
vehicleType                                   NaN                         NaN  
yearOfRegistration                           2017                        2000  
gearbox                                   manuell                         NaN  
powerPS                                        90                           0  
model                                         NaN                         NaN  
odometer                                150,000km                   150,000km  
monthOfRegistration                             4                           0  
fuelType                                   benzin                      benzin  
brand                                  volkswagen               me

### Converting Numeric Data Stored as Text

- The `price` and `odometer` columns have been converted from strings to numbers. 
- The `odometer` has been renamed to `odometer_km` to preserve the unit of measurement.

In [7]:
autos['price'] = autos['price'].str.replace(',','').str.strip('$').astype('int64')
autos['odometer'] = autos['odometer'].str.replace(',','').str.strip('km').astype('int64')

autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)
autos_sample

7                    8  \
dateCrawled           2016-03-16 18:55:19  2016-03-22 16:51:34   
name                 Golf_IV_1.9_TDI_90PS           Seat_Arosa   
seller                             privat               privat   
offerType                         Angebot              Angebot   
price                              $1,990                 $250   
abtest                            control                 test   
vehicleType                     limousine                  NaN   
yearOfRegistration                   1998                 2000   
gearbox                           manuell              manuell   
powerPS                                90                    0   
model                                golf                arosa   
odometer                        150,000km            150,000km   
monthOfRegistration                    12                   10   
fuelType                           diesel                  NaN   
brand                          volkswagen                 seat   
notRepairedDamage                    nein                 nein   
dateCreated           2016-03-16 00:00:00  2016-03-22 00:00:00   
nrOfPictures                            0                    0   
postalCode                          53474                 7426   
lastSeen              2016-04-07 03:17:32  2016-03-26 18:18:10   

                                                             9  \
dateCrawled                                2016-03-16 13:47:02   
name                 Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
seller                                                  privat   
offerType                                              Angebot   
price                                                     $590   
abtest                                                 control   
vehicleType                                                bus   
yearOfRegistration                                        1997   
gearbox                                                manuell   
powerPS                                                     90   
model                                                   megane   
odometer                                             150,000km   
monthOfRegistration                                          7   
fuelType                                                benzin   
brand                                                  renault   
notRepairedDamage                                         nein   
dateCreated                                2016-03-16 00:00:00   
nrOfPictures                                                 0   
postalCode                                               15749   
lastSeen                                   2016-04-06 10:46:35   

                                               10                          11  
dateCrawled                   2016-03-15 01:41:36         2016-03-16 18:45:34  
name                 VW_Golf_Tuning_in_siber/grau  Mercedes_A140_Motorschaden  
seller                                     privat                      privat  
offerType                                 Angebot                     Angebot  
price                                        $999                        $350  
abtest                                       test                     control  
vehicleType                                   NaN                         NaN  
yearOfRegistration                           2017                        2000  
gearbox                                   manuell                         NaN  
powerPS                                        90                           0  
model                                         NaN                         NaN  
odometer                                150,000km                   150,000km  
monthOfRegistration                             4                           0  
fuelType                                   benzin                      benzin  
brand                                  volkswagen               me

## Data Exploration

### Exploring the Odometer and Price Columns

To continue with the data exploration, it's helpful to look at outliers, starting with the `odometer_km` and `price` columns. To do this, it's useful to look at minimum and maximum values, and spot for any values that look unrealistically high or low.

In [8]:
od = autos['odometer_km']
odv = od.value_counts()

od.unique().shape

(13,)

In [9]:
odv.sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

Values in `odometer_km` are much less dispersed, so there is no need to identify and remove outliers.

In [10]:
ap = autos['price']
apv = ap.value_counts()

ap.unique().shape

(2357,)

In [11]:
ap.describe().apply(lambda x: format(x, 'f'))

count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price, dtype: object

In [12]:
apv.sort_index(ascending=False).head(11)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
Name: price, dtype: int64

In [13]:
apv.sort_index(ascending=True).head(11)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
Name: price, dtype: int64

In [14]:
(sum(apv.sort_index(ascending=False).head(10))+sum(apv.sort_index(ascending=True).head(10)))/50000

0.03222

In [15]:
ap[ap.between(13,350000)].describe()

count     48389.000000
mean       5910.348158
std        9069.343912
min          13.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

Based on the sort indices, removing the top and bottom 10 outliers in `price` results to a significant decrease in the standard deviation — from 481,104 to just 9,069. This includes removing posts with prices above a 4.0M dollar mark on one extreme, and those priced below 13.0 dollars — including 0.0-priced items — on the other.

### Exploring the Date Columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. It is easy to differentiate the source by referring to the data dictionary:

| COLUMN NAME | DESCRIPTION |
| :-----------| :-----------|
| **date_crawled** | added by the crawler |
| **last_seen** | added by the crawler |
| **ad_created** | from the website |
| **registration_month** | from the website |
| **registration_year** | from the website |

In [16]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively.

In [17]:
crw_m = autos['date_crawled'].str[:7].value_counts(normalize=True, dropna=False)
lts_m = autos['last_seen'].str[:7].value_counts(normalize=True, dropna=False)
adc_m = autos['ad_created'].str[:7].value_counts(normalize=True, dropna=False)
adc_d = autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False)

In [18]:
crw_m.sort_index()

2016-03    0.8379
2016-04    0.1621
Name: date_crawled, dtype: float64

In [19]:
lts_m.sort_index()

2016-03    0.42582
2016-04    0.57418
Name: last_seen, dtype: float64

In [20]:
adc_m.sort_index()

2015-06    0.00002
2015-08    0.00002
2015-09    0.00002
2015-11    0.00002
2015-12    0.00004
2016-01    0.00024
2016-02    0.00126
2016-03    0.83732
2016-04    0.16106
Name: ad_created, dtype: float64

In [21]:
adc_d.sort_index()[35:50]

2016-02-27    0.00012
2016-02-28    0.00020
2016-02-29    0.00016
2016-03-01    0.00010
2016-03-02    0.00010
2016-03-03    0.00086
2016-03-04    0.00144
2016-03-05    0.02304
2016-03-06    0.01512
2016-03-07    0.03474
2016-03-08    0.03334
2016-03-09    0.03324
2016-03-10    0.03186
2016-03-11    0.03278
2016-03-12    0.03662
Name: ad_created, dtype: float64

Frequency distributions in `autos['date_crawled']`, `autos['last_seen']`, and `autos['ad_created']` seem to yeild expectable results.

Values in both `autos['date_crawled']` and `autos['last_seen']` concentrate between March '16, and the first few days of April. This coincides with the frequencies in `autos['ad_created']`. There is a significant concentration of posts created in March and April '16, which makes up about 99% of the distribution.

The steep incline between February and March '16 may signify: (1) that the turnaround time to a car being sold is about a month; and/or (2) those that didn't sell within a month gets taken down or maybe reposted.

In [22]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration is that the `registration_year` column contains some odd values:
- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

The next step is to work through a few test cases to investigate deeper what's going on with the data.
- Case 1: Registration years before 1906, which was the start of vehicle registration in Germany

In [23]:
autos[autos['registration_year'] <= 1906].iloc[:,[1,2,3,4,5,8,9]].sort_values(by='registration_year')

name  price   abtest  \
22316             VW_Kaefer.__Zwei_zum_Preis_von_einem.   1500  control   
49283                                        Citroen_HY   7750  control   
24511                       Trabant__wartburg__Ostalgie    490  control   
35238  Suche_Skoda_Fabia____Skoda_Fabia_Combi_mit_Klima      0  control   
10556                                        UNFAL_Auto    450  control   
32585                                        UNFAL_Auto    450  control   

      vehicle_type  registration_year   model  odometer_km  
22316          NaN               1000  kaefer         5000  
49283          NaN               1001  andere         5000  
24511          NaN               1111     NaN         5000  
35238          NaN               1500     NaN         5000  
10556          NaN               1800     NaN         5000  
32585          NaN               1800     NaN         5000

- Case 2: Registration years after 2019, which is already 3 years ahead of crawl year

In [24]:
autos[autos['registration_year'] > 2019].iloc[:,[1,2,3,4,5,8,9]].sort_values(by='registration_year')

name  price   abtest  \
27578                                   VW_GOLF_2_SYNCRO   1700     test   
4549                                          Kompressor   1600     test   
453                                           Armee_Jeep   9800     test   
42079                      APE_50___deutsche_Papiere_!!!    500  control   
24519  4x_Winterreifen_auf_Alufelge_der_naechste_Wint...    250     test   
49153                                        Corsa_c20xe   2500     test   
22799                                  Subaru_Impreza_GT   9000     test   
4164   Verkaufe_DESIGN_Streifen_/_Aufkleber_VW__Opel_...     49  control   
27618  Golf_1_75ps_5911km_Original_Automatik_einer_de...   7999     test   
8360                            Vito_touret_119_Blue_Tec  42800  control   
25003              Reo_Vorkriegs_Oldtimer_Rennwagen_1928   6500  control   
13559          Saab_9000_CSE_Automatik_2_3_ltr._mit_EGSD      0  control   
49910  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...  22200     test   
6308                               Kaufe_Autos_jeglicher      0     test   
14341                                  Hole_kostenlos_ab      0     test   
8012                       Opel_GT_Karosserie_mit_Brief!    700     test   
33950  58er_karmann_ghia_lowlight_Kaefer__zum_restaur...   7999     test   
38076                                 Mercedes_Benz_A180  18000     test   

      vehicle_type  registration_year     model  odometer_km  
27578          NaN               2800       NaN         5000  
4549           NaN               4100       NaN         5000  
453            NaN               4500    andere         5000  
42079          NaN               4800       NaN         5000  
24519          NaN               5000    andere         5000  
49153          NaN               5000     corsa         5000  
22799          NaN               5000   impreza         5000  
4164           NaN               5000      golf         5000  
27618          NaN               5911      golf        10000  
8360           NaN               6200      vito        10000  
25003          NaN               8888       NaN        10000  
13559          NaN               9000      9000       150000  
49910          NaN               9000    andere        10000  
6308           NaN               9996       NaN        10000  
14341          NaN               9999       NaN        10000  
8012           NaN               9999    andere        10000  
33950          NaN               9999    kaefer        10000  
38076          NaN               9999  a_klasse        10000

Based on the data, posts with registration years before 1906 and after 2019 logically have incorrect registration years posted. Most likely these posts need to be removed or updated.

In [25]:
autos = autos[autos['registration_year'].between(1906,2019)]

autos.loc[:,'registration_year'].round(-1).value_counts().sort_index()

1910        9
1930        5
1940        9
1950       10
1960      102
1970      251
1980      532
1990     2721
2000    27307
2010    15367
2020     3663
Name: registration_year, dtype: int64

Looking at the registration year distribution after the outliers have been removed, it looks clear that the majority of the registration years fall within the last couple of decades.

### Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. Aggregation can be used to understand the brand column.

In [31]:
autos[7:12].T

7                    8   \
date_crawled         2016-03-16 18:55:19  2016-03-22 16:51:34   
name                Golf_IV_1.9_TDI_90PS           Seat_Arosa   
price                               1990                  250   
abtest                           control                 test   
vehicle_type                   limousine                  NaN   
registration_year                   1998                 2000   
gearbox                          manuell              manuell   
power_ps                              90                    0   
model                               golf                arosa   
odometer_km                       150000               150000   
registration_month                    12                   10   
fuel_type                         diesel                  NaN   
brand                         volkswagen                 seat   
unrepaired_damage                   nein                 nein   
ad_created           2016-03-16 00:00:00  2016-03-22 00:00:00   
postal_code                        53474                 7426   
last_seen            2016-04-07 03:17:32  2016-03-26 18:18:10   

                                                           9   \
date_crawled                              2016-03-16 13:47:02   
name                Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
price                                                     590   
abtest                                                control   
vehicle_type                                              bus   
registration_year                                        1997   
gearbox                                               manuell   
power_ps                                                   90   
model                                                  megane   
odometer_km                                            150000   
registration_month                                          7   
fuel_type                                              benzin   
brand                                                 renault   
unrepaired_damage                                        nein   
ad_created                                2016-03-16 00:00:00   
postal_code                                             15749   
last_seen                                 2016-04-06 10:46:35   

                                              10                          11  
date_crawled                 2016-03-15 01:41:36         2016-03-16 18:45:34  
name                VW_Golf_Tuning_in_siber/grau  Mercedes_A140_Motorschaden  
price                                        999                         350  
abtest                                      test                     control  
vehicle_type                                 NaN                         NaN  
registration_year                           2017                        2000  
gearbox                                  manuell                         NaN  
power_ps                                      90                           0  
model                                        NaN                         NaN  
odometer_km                               150000                      150000  
registration_month                             4                           0  
fuel_type                                 benzin                      benzin  
brand                                 volkswagen               mercedes_benz  
unrepaired_damage                           nein                         NaN  
ad_created                   2016-03-14 00:00:00         2016-03-16 00:00:00  
postal_code                                86157                       17498  
last_seen                    2016-04-07 03:16:21         2016-03-16 18:45:34

In [43]:
brands_mp = {}
brands_km = {}

for b in autos['brand'].unique():
    mean_price = autos[autos['brand'] == b].loc[:,'price'].mean()
    mean_km = autos[autos['brand'] == b].loc[:,'odometer_km'].mean()
    brands_mp[b] = round(mean_price, 2)
    brands_km[b] = round(mean_km, 2)
    
autos_new = pd.DataFrame({'mean_price': pd.Series(brands_mp), 'mean_mileage': pd.Series(brands_km)})

autos_new[:10].T

peugeot        bmw  volkswagen      smart       ford  \
mean_price      3010.87    8254.44     6385.35    3482.97    7105.66   
mean_mileage  127352.34  132544.22   129012.36  100756.06  124131.93   

               chrysler       seat    renault  mercedes_benz       audi  
mean_price      3286.06    4223.65    2351.30       29511.58    8965.56  
mean_mileage  133149.17  122186.17  128223.79      130937.24  129643.94

In [44]:
autos_new[10:20].T

sonstige_autos       opel      mazda   porsche      mini  \
mean_price          38567.64    5104.25    3962.54  44537.98  10392.39   
mean_mileage        87776.75  129365.15  125132.10  97363.95  89375.00   

                 toyota     dacia     nissan       jeep       saab  
mean_price      5097.94   5897.74    4588.88   11377.55    3183.49  
mean_mileage  115988.65  84728.68  118978.78  127522.94  143670.89

## Conclusion

In this guided project, a variety of pandas methods have been used to explore and understand a data set on car listings. Here are some next steps for you to consider:

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
- How much cheaper are cars with damage than their non-damaged counterparts?